---
title: Differential Digestion
author: Soren Heidelbach
output:
  bookdown::html_document2:
    code_folding: hide
    toc: true
    toc_depth: 2
  
---
```{r, setup}
setwd("/home/ubuntu/vol_store/differential_digestion")
pacman::p_load(
  "data.table",
  "wrapr",
  "ggplot2",
  "magrittr",
  "ggside",
  "gt",
  "gtExtras",
  "stringr",
  "scatterPlotMatrix",
  "GGally",
  "knitr",
  "kableExtra",
  "bookdown",
  "patchwork",
  "metR" # only for discretizing cotinous color scale
)
knitr::opts_knit$set(
  root.dir = "/home/ubuntu/vol_store/differential_digestion"
)
knitr::opts_chunk$set(
  fig.align = "center",
  fig.width = 8,
  fig.height = 6,
  message = FALSE,
  warning = FALSE,
  echo = TRUE,
  fig.path = "analysis/figures",
  out.width = "70%"
)
# General theme elements for all plots
plot_theme <- theme_bw(
  base_size = 12,
  base_family = "Helvetica"
)
contig_size_scale  <- scale_size_continuous(
  range = c(1, 7),
  breaks = c(1e4, 1e5, 1e6, 1e7)
)
```

# Differential Digestion
This report aims to investigate the potential in digesting DNA extraction with methylation sensitive enzyme in order to create taxonomic Differential coverage.


```{r load_data}
# Load bin stats from mmlong2
bins <- fread("mmlong2/results/mmlong2_bins.tsv")[
  , quality := ifelse(Completeness >= 90 & Contamination <= 10, "HQ", "MQ")
  ][
    , Type := "mmlong2"
  ][
    , GC_Content := GC_Content * 100
  ]

# Load coverage ID to sample name
coverage_sample_id <- fread("mmlong2/tmp/binning/reads.csv", header = FALSE)[
    , .(path = V2)
  ][
    , sample := str_extract(path, "[^/]*$")
  ][
    , sample := str_replace(sample, ".fastq", "")
  ][
    , coverage_id := as.character(seq_len(.N))
  ][
    , !"path"
  ]

# Load digestion coverage profile
coverage_digestion <- fread("mmlong2/results/mapping/NP_MboI_R1041.cov.tsv")[
    , .(contig = `#rname`, NP_MboI_R1041_mean_depth = meandepth)
  ]

# Load coverage profiles
coverage <- {
  coverage <- fread("mmlong2/tmp/binning/metabat_cov.tsv") %>%
    melt(id.vars = c("contigName", "contigLen", "totalAvgDepth")) %>%
    setnames(c("contig", "length", "mean_depth_all", "sample_and_type", "value"))
  coverage[
      , coverage_id := str_extract(sample_and_type, "^[^_]*")
    ][
      , value_type := fifelse(
          str_detect(sample_and_type, "var"),
          "var_depth",
          "mean_depth"
        )
    ][
      , !"sample_and_type"
    ]
  # Merge coverage profiles with sample coverage IDs
  coverage[coverage_sample_id, on = "coverage_id"] %>%
    dcast(
      contig + length + mean_depth_all ~ sample + value_type,
      value.var = "value"
    )
}

# Load contig to bin mapping
contig_to_bin <- fread("mmlong2/tmp/binning/contig_bin.tsv", header = FALSE) %>%
  setnames(c("contig", "bin"))

# Add bin info to contigs
contigs <- Reduce(
  function(x, y) {
    shared_columns  <- intersect(names(x), names(y))
    message(paste0(
      "Merging by ",
      paste(shared_columns, collapse = ", ")
    ))
    merge(x, y, by = shared_columns, all = TRUE)
  },
  list(coverage, coverage_digestion, contig_to_bin, bins)
)
contigs <- contigs[
    , bin_num := as.numeric(str_extract(bin, "\\d+$"))
  ][
    length > 1e4,
  ]

```

## Approach
The constitue an anaerobic digfester DNA extract from 2017. The sample was sequenced directly with Nanopore R10.4.1 (400bps), and a MboI digestion of the sample was sequenced, using the same chemistry. 
In addityion to these two sequencing runs several other sequencing were avalible, from Illumina, PacBio  and Nanopore with different chemistries. 
These were used to compare the coverage profiles of the different samples (`r (coverage_sample_id$sample[!grepl("Mbo|400bps", coverage_sample_id$sample)])`).

All of the data was used in mmlin2-lite workflow. The The R10.4.1 (400bps) sample was used for assembly and all but MboI was used for coverage profiles. 
MboI was not inputted into the worklfow, but mapped seperately. 

The ressults of the workflopw i presented in Table \@ref(tab:binQualityTable). A total of `r nrow(bins)` MAGs were generated, of which `r sum(bins$quality == "HQ")` were high quality.

```{r binQualityTable, echo = FALSE}

bins[
  , .(
    Total = .N,
    HQ = sum(quality == "HQ"),
    MQ = sum(quality == "MQ"),
    mean_genom_size_mb = mean(Genome_Size)/1e6,
    mean_n50_kb = mean(Contig_N50)/1e3
    ), by = c("Type")
  ] %>%
  `[`(
    , Total := HQ + MQ
  ) %>%
  kbl(
    caption = "Number of bins generated",
    digits = 2,
    booktabs = TRUE,
    col.names = c("Type", "Total MAGs", "HQ-MAGs", "MQ-MAGs", "Mean genome size (Mb)", "Mean N50 (kb)")
  ) %>%
 kable_styling()
```

Three sequencing runs were initially used to compare the different coverage profiles; The undigested Nanopore R10.4.1 (400bps) sample, the MboI digested Nanopore R10.4.1 (400bps) sample and the Illumina 2020-11-04 sample.
The comparisons are shown in figure \@ref(fig:coverageProfiles). The comparison between dighested and undigested NAnopore samples show that coverage is generally only depleted going from undighested to digested sample. 
Additionally, contigs of the same bins seem to have equally depleted coverage. 


The coverage profile of the MboI digestion was compared to the R10.4.1 (400bps) sample, as these two are from the same DNA extraction and therefore resemble each other the most.
The difference is shown in Figure \@ref(fig:coverageProfiles). 



```{r coverageProfiles, fig.asp=0.9, out.width = "100%", fig.cap="Difference in coverage profiles of Nanopore MboI digestion, Nanopore Untreated and Illumina 2020-11-04. (1) Undigested sample vs. Illumina 2020-11-04. (2) Undigested sample vs. MboI digestion. (3) Coverage ratio of undigested to MboI digestion. (4) Density plot of the coverage ratio. Equal coverage is marked with a red line. (3 & 4) Dotted lines indicate different coverage ratios."}
shared_ggplot_objects  <- function() {
  list(
    geom_hline(yintercept = 1, linetype = 1, color = "red", size = 2),
    geom_text(x = -1, y = -1.05, label = "Untreated : MboI", size = 3, vjust=0, hjust=0),
    geom_text(x = -1, y = 0.05, label = "1 : 1", size = 3, vjust=0, hjust=0),
    geom_text(x = -1, y = 1.05, label = "10 : 1", size = 3, vjust=0, hjust=0),
    geom_hline(yintercept = 10, linetype = 2),
    geom_text(x = -1, y = 2.05, label = "100 : 1", size = 3, vjust=0, hjust=0),
    geom_hline(yintercept = 100, linetype = 2),
    geom_text(x = -1, y = 3.05, label = "1000 : 1", size = 3, vjust=0, hjust=0),
    geom_hline(yintercept = 1000, linetype = 2),
    plot_theme,
    scale_y_log10(limits = c(0.1, NA))
  )
}

cov_vs_cov_plot <- function() {
  list(
    geom_abline(slope = 1, intercept = 0, color = "red", size = 2),
    geom_hline(yintercept = 1, color = "black"),
    geom_vline(xintercept = 1, color = "black"),
    geom_point(alpha = 0.2, shape = 21),
    plot_theme,
    scale_x_log10(limits = c(0.1, 1e3), expand = c(0, 0)),
    scale_y_log10(limits = c(0.1, 1e3), expand = c(0, 0)),
    contig_size_scale,
    scale_fill_viridis_c()
  )
}

il_vs_np <- ggplot(contigs) +
  aes(
    y = NP_R1041_400bps_sub_mean_depth,
    x = `IL-201104_mean_depth`,
    size = length,
    fill = bin_num
  ) +
  cov_vs_cov_plot() +
  labs(
    y = "Contig Coverage\nNanopore Untreated",
    x = "Contig Coverage\nIllumina 2020-11-04",
    size = "Contig \nlength (bp)",
    fill = "Bin number"
  )

mboi_vs_np  <- ggplot(contigs) +
  aes(
    y = NP_R1041_400bps_sub_mean_depth,
    x = NP_MboI_R1041_mean_depth,
    size = length,
    fill = bin_num
  ) +
  cov_vs_cov_plot() +
  labs(
    y = "Contig Coverage\nNanopore Untreated",
    x = "Contig Coverage\nNanopore MboI digest",
    size = "Contig \nlength (bp)",
    fill = "Bin number"
  )  +
    theme(
      axis.text.y = element_blank(),
      axis.ticks.y = element_blank(),
      axis.title.y = element_blank()
    )


enrichment_plot  <- contigs[
    NP_R1041_400bps_sub_mean_depth != 0,
    coverage_fold_change := NP_R1041_400bps_sub_mean_depth / NP_MboI_R1041_mean_depth
  ]  %>%
  ggplot() +
    aes(
      x = NP_R1041_400bps_sub_mean_depth,
      y = coverage_fold_change,
      size = length,
      fill = bin_num
    )  +
    shared_ggplot_objects() +
    geom_point(alpha = 0.2, shape = 21)  +
    contig_size_scale +
    scale_alpha_continuous(range = c(0.1, 1)) +
    labs(
      x = "Contig Coverage\nNanopore Untreated",
      y = "Coverage ratio \n(Untreated / MboI)",
      size = "Contig \nlength (bp)",
      fill = "Bin number"
    ) +
    scale_x_log10(limits = c(0.1, NA)) +
    scale_fill_viridis_c()

density_plot <- contigs[
    NP_R1041_400bps_sub_mean_depth != 0,
    coverage_fold_change := NP_R1041_400bps_sub_mean_depth / NP_MboI_R1041_mean_depth
  ]  %>%
  ggplot() +
    aes(y = coverage_fold_change) +
    geom_density(aes(weight = length), fill = "gray") +
    theme_bw() +
    shared_ggplot_objects() +
    labs(
      y = "Coverage ratio \n(Untreated / MboI)",
      x = "Density \n(weighted  with contig length)"
    ) +
    theme(
      axis.text.y = element_blank(),
      axis.ticks.y = element_blank(),
      axis.title.y = element_blank()
    )

((il_vs_np + mboi_vs_np) / (enrichment_plot + density_plot)) +
  plot_layout(guides = "collect") +
  plot_annotation(tag_levels = "1", tag_prefix = "")
```


```{r}
coverage_columns <- names(contigs)[grepl("mean_depth", names(contigs))]
plot_cols <- coverage_columns[base::grepl(
  pattern = paste(
    c(
      "NP_R1041_400bps",
      "NP_R941",
      "MboI",
      "IL-201702",
      "IL_MiSeq",
      "PB_HiFi"
    ),
    collapse = "|"
  ),
  x = coverage_columns
)]
plot_data <- cbind(
  {
    contigs[
        , ..plot_cols
      ][
        , lapply(.SD, log10)
      ] %>%
      setnames(
        paste0(
          str_remove_all(names(.), "NP_|_mean_depth|_sub"),
          " (log10)"
        )
      )
  },
  contigs[, c("length", "GC_Content")]
)
plot_data[plot_data == -Inf] <- NA
plot_data_no_na <- plot_data[complete.cases(plot_data), ]
```

```{r}
# Matrix plot
upper_density <- function(data, mapping, ...) {
  ggplot(data = data, mapping = mapping) +
    geom_density2d_filled(mapping = aes(fill = after_stat(level))) +
    scale_fill_discretised(low = "white", high = "#000000")
}
lower_scatter <- function(data, mapping, ...) {
  ggplot(data = data, mapping = mapping) +
    geom_smooth(method = "lm", col = "red") +
    geom_point(mapping = aes(size = length), alpha = 0.1, shape = 16) +
    scale_size_continuous(trans = "log10", range = c(0.1, 1))
}
diag_density <- function(data, mapping, ...) {
  ggplot(data = data, mapping = mapping) +
    geom_density(aes(weight = length), fill = "gray")
}
ggpairs(
    plot_data_no_na,
    columns = names(plot_data)[!grepl("length", names(plot_data))],
    lower = list(continuous = lower_scatter),
    diag = list(continuous = diag_density)
  ) +
  theme_bw()
```



```{r}




# Check 
contigs[, .(mean_depth_all = mean(mean_depth_all)), by = bin] %>%
  ggplot(aes(x = mean_depth_all)) +
  geom_histogram(bins = 20) +
  plot_theme +
  scale_x_log10()
```  



```{r}
contigs_binned  <- contigs[!is.na(bin_num), ]
anova_within_bins  <- lapply(
  coverage_columns,
  function(x) {
    result  <- lm(
      contigs_binned$bin_num ~ contigs_binned[[x]]
    )  %>%
      anova()  %>%
      as.data.frame()
    result$cov_profile <- x
    rownames(result) <- c(x, "Residuals")
    result[1, ]
  }
) %>%
  rbindlist() %>%
  kbl(
    caption = "ANOVA results for each coverage profile ()",
    digits = 2,
    booktabs = TRUE
  ) %>%
  kable_styling()

for (x in coverage_columns) {
  aov(
      contigs[[x]] ~ contigs$bin_num
    )
}
    aov(
      contigs[[x]] ~ contigs$bin_num
    )
aov(
  NP_R1041_400bps_sub_mean_depth ~ bin_num,
  data = contigs
) %>%
  summary()
aov(
  NP_MboI_R1041_mean_depth ~ bin_num,
  data = contigs
) %>%
  summary()
```
```






```{r, out.width="100%"}
x_order <- contigs[
  , .(mean_depth_all = mean(log(mean_depth_all))), by = bin_num
  ][
    order(mean_depth_all, decreasing = TRUE),
    as.character(bin_num)
  ]
melt(
  contigs,
  id.vars = c("bin_num", "length"),
  measure.vars = coverage_columns[c(14, 17)]
)  %>% 
  `[`(
    , variable := str_remove_all(variable, "NP_|_mean_depth|_sub")
  )  %>% 
  ggplot() +
    aes(x = as.character(bin_num), size = length, color = variable, fill = !is.na(bin_num)) +
    aes(y = value) +
    geom_boxplot(outlier.alpha = 0) +
    contig_size_scale +
    scale_fill_manual(values = c("red", "black")) +
    scale_color_manual(values = c("#007a00", "#ca37dd")) +
    scale_y_log10(limits = c(1e-2, 1e3)) +
    scale_x_discrete(limits = x_order) +
    plot_theme +
    theme(
      legend.position = "bottom",
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank()
    ) +
    labs(
      x = "Bin number",
      y = "Coverage (log10)",
      fill = "Contig mapped",
      color = "Coverage \nprofile"
    )


undigest_bins_cov <- ggplot(contigs) +
  aes(x = as.character(bin_num), size = length, color = is.na(bin_num)) +
  aes(y = NP_R1041_400bps_sub_mean_depth) +
  geom_boxplot(outlier.alpha = 0) +
  contig_size_scale +
  scale_color_manual(values = c("black", "red")) +
  scale_y_log10(limits = c(1e-2, 1e3)) +
  scale_x_discrete(limits = x_order) +
  plot_theme
digest_bin_cov  <- ggplot(contigs) +
  aes(x = as.character(bin_num), size = length, color = is.na(bin_num)) +
  aes(y = NP_MboI_R1041_mean_depth) +
  geom_boxplot(outlier.size = 0.5) +
  contig_size_scale +
  scale_color_manual(values = c("black", "red")) +
  scale_y_log10(limits = c(1e-2, 1e3)) +
  scale_x_discrete(limits = x_order) +
  plot_theme +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank()
  )

  digest_bin_cov / undigest_bins_cov
```


```{r}
median_difference  <- contigs[
  , .(
    mean_digest = median(NP_MboI_R1041_mean_depth),
    mean_nondigest = median(NP_R1041_400bps_sub_mean_depth)
    ), by = bin_num
][
  , diff := mean_nondigest / mean_digest
]  
median_difference %>%
  ggplot() +
    geom_hline(yintercept = 1, col = "black", lty = 1) +
    geom_point(aes(x = as.character(bin_num), y = mean_nondigest), size = 1) +
    geom_segment(
      aes(
        x = as.character(bin_num),
        y = mean_nondigest,
        xend = as.character(bin_num),
        yend = mean_digest,
        col = diff),
      arrow = arrow(length = unit(0.2, "cm"), type = "closed")
    ) +
    scale_y_log10() +
    scale_x_discrete(limits = x_order) +
    scale_color_gradient(
      high = "#ff2929",
      low = "#2dceff",
      na.value = "black",
      trans = "log10",
      breaks = c(1, 3, 10, 30, 100),
      limits = c(0.5, 100)
    ) +
    plot_theme +
    theme(
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank()
    ) +
    labs(
      title = "Coverage difference between digest and nondigest within bins",
      x = "Bin number",
      y = "Median Bin Coverage",
      col = "Nondigest / Digest"
    ) 

    
```


```{r}
median_difference[ 1 - 1/diff > 0.8, .N] / nrow(median_difference)
```




```{r}

motif_frequency  <- function(motif, gc) {
  n_at  <- nchar(str_remove_all(motif, "G|C"))
  n_gc  <- nchar(str_remove_all(motif, "A|T"))

  prob_gc  <- gc / 2
  prob_at  <- (1 - gc) / 2

  prob_motif  <- prob_gc^n_gc * prob_at^n_at
  motif_freq  <- 1 / prob_motif
  motif_freq
}
GCs <- seq(0.1, 0.9, 0.1)
freq  <- lapply(
  GCs,
  function(x) motif_frequency("GGATCA", x)
)  %>%
  unlist() %>%
  round(0)
names(freq) <- GCs
freq

```


```{r chunkedCoverage}




```